mesonbuild/backend/ninjabackend.py

```diff
@@ -1131,9 +1131,14 @@ def generate_vala_compile(self, target, outfile):
             # file is outside the build directory, the path components will be
             # stripped and just the basename will be used.
             if isinstance(gensrc, (build.CustomTarget, build.GeneratedList)) or gensrc.is_built:
-                vala_c_file = os.path.splitext(vala_file)[0] + '.c'
-            else:
                 vala_c_file = os.path.splitext(os.path.basename(vala_file))[0] + '.c'
```

Comment:
You've inverted the if condition here, so now we will see collisions with the basename of generated files.

Reply:
There are collisions, I saw that when I removed the `-b` (`--basedir`) option for `valac`. It just overwrites the existing test.c file. That why `--basedir` is needed, so `valac` outputs the C with the directory name.

Reply:
I see, my bad. I get what you're doing here now.